## **2.1. Data Manipulation**

In [180]:
!python -m pip install --upgrade pip
!pip3 uninstall d2l -y
!pip3 uninstall tensorflow -y
!pip3 uninstall numpy -y
!pip3 uninstall matplotlib -y
!pip3 install d2l
!pip3 install tensorflow
!pip3 install pandas==1.3.5
!pip3 install matplotlib

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Found existing installation: d2l 0.17.6
Uninstalling d2l-0.17.6:
  Successfully uninstalled d2l-0.17.6
Found existing installation: tensorflow 2.12.0
Uninstalling tensorflow-2.12.0:
  Successfully uninstalled tensorflow-2.12.0
Found existing installation: numpy 1.23.5
Uninstalling numpy-1.23.5:
  Successfully uninstalled numpy-1.23.5
Found existing installation: matplotlib 3.5.1
Uninstalling matplotlib-3.5.1:
  Would remove:
    /usr/local/lib/python3.10/dist-packages/matplotlib-3.5.1-py3.10-nspkg.pth
    /usr/local/lib/python3.10/dist-packages/matplotlib-3.5.1.dist-info/*
    /usr/local/lib/python3.10/dist-packages/matplotlib/*
    /usr/local/lib/python3.10/dist-packages/mpl_toolkits/axes_grid/*
    /usr/local/lib/python3.10/dist-packages/mpl_toolkits/axes_grid1/*
    /usr/local/lib/python3.10/dist-packages/mpl_toolkits/axisartist/*
    /usr/local/lib/python3.10/dist-packages/mpl_toolkit

In [270]:
%matplotlib inline
import tensorflow as tf

#### 2.1.1. Getting Started

기본적인 Tensor의 생성 및 Reshaping에 관한 인터페이스를 살펴본다.
본인은 주로 Tensorflow Framework를 사용하기 때문에 Tensorflow 버전 코드로 실습하였다.

In [271]:
x = tf.range(12, dtype=tf.float32)
x

<tf.Tensor: shape=(12,), dtype=float32, numpy=
array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11.],
      dtype=float32)>

위 코드는 range(12) 값들을 가지고 1차원 텐서를 만들었다.
shape는 (12,) 이다.

In [272]:
tf.size(x)

<tf.Tensor: shape=(), dtype=int32, numpy=12>

tf.size(x)로 size를 가져올 수 있다. size 값 자체는 차원이 없는 것을 확인할 수 있다. 위 경우 size는 12이다.

In [273]:
x.shape

TensorShape([12])

x.shape attribute로 shape를 확인하는 코드이다. 위 경우 (12,)인 것을 확인할 수 있다. 이 attribute는 reshape 메소드로 조작할 수 있다.

In [274]:
X = tf.reshape(x, (3, 4))
X

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]], dtype=float32)>

위와 같이 x를 (3,4)로 reshape한 결과를 X에 넣었다.

In [275]:
X = tf.reshape(x, (-1, 4))
X

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]], dtype=float32)>

In [276]:
X = tf.reshape(x, (3, -1))
X

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]], dtype=float32)>

reshape를 할 때 위와 같이 -1을 넣어서 자동으로 추론하여 reshape를 하게끔 할 수 있다. -1 값은 하나만 있어야 한다. 당연한 게, 두 개 이상 있으면 여러 경우의 수가 가능하기 때문이다.

In [277]:
tf.zeros((2,3,4))

<tf.Tensor: shape=(2, 3, 4), dtype=float32, numpy=
array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]], dtype=float32)>

zeros를 통해 모든 값이 0인 텐서를 생성할 수 있다.

In [278]:
tf.ones((2,3,4))

<tf.Tensor: shape=(2, 3, 4), dtype=float32, numpy=
array([[[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]],

       [[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]]], dtype=float32)>

ones를 통해 비슷하게 모든 값이 1인 텐서를 생성할 수 있다.

In [279]:
tf.random.normal(shape=[3,4])

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 1.8133974 ,  0.9544408 , -0.12012713, -0.7727535 ],
       [ 0.13971083, -0.76147014,  0.48794216, -0.18504989],
       [ 0.25194666,  0.34384906, -0.50397193, -2.0321836 ]],
      dtype=float32)>

random.normal로 원소들이 표준 정규 분포를 따르는 텐서를 만들 수도 있다.

In [280]:
tf.constant([[2,1,4,3],[1,2,3,4],[4,3,2,1]])

<tf.Tensor: shape=(3, 4), dtype=int32, numpy=
array([[2, 1, 4, 3],
       [1, 2, 3, 4],
       [4, 3, 2, 1]], dtype=int32)>

constant를 통해 list로 직접 값을 전달하여 텐서를 만들 수 있다. 위 경우 nested list로 전달하였는데, 바깥 리스트는 axis 0에 해당하고, 안쪽 리스트는 axis 1에 해당한다. 예전에 Numpy를 공부하였을 때, axis라는 것은 두 array를 붙이거나 할 때 기준이 되는 축을 의미하는 것으로 이해하였다.

#### 2.1.2. Indexing and Slicing

In [281]:
X, X[-1], X[1:3]

(<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
 array([[ 0.,  1.,  2.,  3.],
        [ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.]], dtype=float32)>,
 <tf.Tensor: shape=(4,), dtype=float32, numpy=array([ 8.,  9., 10., 11.], dtype=float32)>,
 <tf.Tensor: shape=(2, 4), dtype=float32, numpy=
 array([[ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.]], dtype=float32)>)

List의 그것과 비슷하게 Tensor도 Indexing과 Slicing이 가능하다.

In [282]:
X_var = tf.Variable(X)
X_var[1,2].assign(9)
X_var

<tf.Variable 'Variable:0' shape=(3, 4) dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  9.,  7.],
       [ 8.,  9., 10., 11.]], dtype=float32)>

TensorFlow의 Tensor는 constant하다. 즉 값을 바꿀 수가 없기 때문에, Variable이라는 다른 container를 사용해야 한다. PyTorch같은 경우는 tensor 안에서 그대로 값을 바꿀 수 있는 것 같다.

In [283]:
X_var = tf.Variable(X)
X_var[:2, :].assign(tf.ones(X_var[:2, :].shape, dtype=tf.float32)*12)

<tf.Variable 'UnreadVariable' shape=(3, 4) dtype=float32, numpy=
array([[12., 12., 12., 12.],
       [12., 12., 12., 12.],
       [ 8.,  9., 10., 11.]], dtype=float32)>

조금 복잡해 보이지만, 첫 번째 행과 두 번째 행을 모두 12로 바꾸는 코드이다. tf.ones로 Slicing한 것과 같은 2*4 Tensor를 만든 뒤 12를 곱해 그것을 할당했다.

#### 2.1.3. Operations

In [284]:
tf.exp(x)

<tf.Tensor: shape=(12,), dtype=float32, numpy=
array([1.0000000e+00, 2.7182817e+00, 7.3890562e+00, 2.0085537e+01,
       5.4598148e+01, 1.4841316e+02, 4.0342877e+02, 1.0966332e+03,
       2.9809580e+03, 8.1030840e+03, 2.2026465e+04, 5.9874141e+04],
      dtype=float32)>

Tensor에 대하여 각 원소마다 연산을 하는 Element-wise Operation을 수행할 수 있다. 위의 경우 x의 원소에 대해 각각 $e^x$의 연산을 수행한 결과이다.

In [285]:
x = tf.constant([1.0, 2, 4, 8])
y = tf.constant([2.0, 2, 2, 2])
x + y, x - y, x * y, x / y, x ** y

(<tf.Tensor: shape=(4,), dtype=float32, numpy=array([ 3.,  4.,  6., 10.], dtype=float32)>,
 <tf.Tensor: shape=(4,), dtype=float32, numpy=array([-1.,  0.,  2.,  6.], dtype=float32)>,
 <tf.Tensor: shape=(4,), dtype=float32, numpy=array([ 2.,  4.,  8., 16.], dtype=float32)>,
 <tf.Tensor: shape=(4,), dtype=float32, numpy=array([0.5, 1. , 2. , 4. ], dtype=float32)>,
 <tf.Tensor: shape=(4,), dtype=float32, numpy=array([ 1.,  4., 16., 64.], dtype=float32)>)

단항 연산 뿐만 아니라 위와 같이 이항 연산도 Element-wise로 수행이 가능하다.

In [286]:
X = tf.reshape(tf.range(12, dtype=tf.float32), (3,4))
Y = tf.constant([[2.0, 1, 4, 3], [1, 2, 3, 4], [4,3,2,1]])
tf.concat([X, Y], axis = 0),  tf.concat([X, Y], axis = 1)

(<tf.Tensor: shape=(6, 4), dtype=float32, numpy=
 array([[ 0.,  1.,  2.,  3.],
        [ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.],
        [ 2.,  1.,  4.,  3.],
        [ 1.,  2.,  3.,  4.],
        [ 4.,  3.,  2.,  1.]], dtype=float32)>,
 <tf.Tensor: shape=(3, 8), dtype=float32, numpy=
 array([[ 0.,  1.,  2.,  3.,  2.,  1.,  4.,  3.],
        [ 4.,  5.,  6.,  7.,  1.,  2.,  3.,  4.],
        [ 8.,  9., 10., 11.,  4.,  3.,  2.,  1.]], dtype=float32)>)

concat으로 두 Tensor를 이어붙일 수 있다. 아까 잠깐 나왔던 axis가 다시 등장했는데, axis가 0이면 바깥 리스트를 기준으로 안 쪽 리스트를 그대로 이어붙였고, axis가 1이면 안쪽 리스트를 기준으로 대응되는 행끼리 이어 붙인 모습이다.

In [287]:
X == Y, X != Y

(<tf.Tensor: shape=(3, 4), dtype=bool, numpy=
 array([[False,  True, False,  True],
        [False, False, False, False],
        [False, False, False, False]])>,
 <tf.Tensor: shape=(3, 4), dtype=bool, numpy=
 array([[ True, False,  True, False],
        [ True,  True,  True,  True],
        [ True,  True,  True,  True]])>)

Element-wise로 논리 연산도 가능하다.

In [288]:
tf.reduce_sum(X)

<tf.Tensor: shape=(), dtype=float32, numpy=66.0>

reduce_sum 메소드를 이용해 모든 원소의 합을 구할 수 있다.

#### 2.1.4. Broadcasting

2.1.3. 절에서 다루었던 Operation들은 모두 같은 shape끼리 연산을 수행하였다. 하지만 특정 조건을 만족한다면 shape가 달라도 연산이 가능하고, 이 메커니즘을 Broadcasting이라 한다.

In [289]:
a = tf.reshape(tf.range(3), (3,1))
b = tf.reshape(tf.range(2), (1,2))

a, b

(<tf.Tensor: shape=(3, 1), dtype=int32, numpy=
 array([[0],
        [1],
        [2]], dtype=int32)>,
 <tf.Tensor: shape=(1, 2), dtype=int32, numpy=array([[0, 1]], dtype=int32)>)

In [290]:
a + b

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[0, 1],
       [1, 2],
       [2, 3]], dtype=int32)>

위 코드에서 a는 $3 \times 1$, b는 $1 \times 2$ Tensor로 모양이 다르지만, $3 \times 2$ 모양의 Tensor가 결과로 나왔다.

#### 2.1.5. Saving Memory

In [291]:
before = id(Y)
Y = Y + X
id(Y) == before

False

연산을 실행하면 메모리 할당이 일어날 수 있다. Python에서는 Y+X를 수행한 결과를 새 메모리에 할당하고 Y가 그 메모리를 pointing하게끔 한다. 이는 메모리 낭비로 이어질 수 있다.

In [292]:
Z = tf.Variable(tf.zeros_like(Y))
print('id(Z): ', id(Z))
Z.assign(X+Y)
print('id(Z): ', id(Z))

id(Z):  140528696652464
id(Z):  140528696652464


위와 같이 Variable 컨테이너를 선언하고 assign을 통해 Z 안에서 연산을 수행하는 방법 (in-place)으로 해결할 수 있다. zeros_like 메소드는 argument와 shape가 같고 0으로 채워진 Tensor를 반환한다.

In [293]:
@tf.function
def computation(X, Y):
  Z = tf.zeros_like(Y)
  A = X + Y
  B = A + Y
  C = B + Y
  return C + Y

computation(X, Y)

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[ 8.,  9., 26., 27.],
       [24., 33., 42., 51.],
       [56., 57., 58., 59.]], dtype=float32)>

위와 같이 @tf.function이라는 decorator를 붙여서 자동으로 최적화를 수행할 수 있다. 이 또한 메모리 낭비를 줄일 수 있는 방법 중 하나이다.

#### 2.1.6. Conversion to Other Python Objects

In [294]:
A = X.numpy()
B = tf.constant(A)
type(A), type(B)

(numpy.ndarray, tensorflow.python.framework.ops.EagerTensor)

위같이 numpy 메소드를 통해 NumPy Tensor로 자료형을 변환할 수 있다.

In [295]:
a = tf.constant([3.5]).numpy()
a, a.item(), float(a), int(a)

(array([3.5], dtype=float32), 3.5, 3.5, 3)

item이라는 Python의 built-in 메소드를 사용하여 a를 스칼라로 변환한 모습이다. 아쉽게도 size가 1인 Tensor에 대해서만 사용이 가능하다.

Tensor는 머신러닝에서 데이터 저장 및 조작을 위한 코어이다. Tensor는 위와 같이 다양한 Operation, Broadcasting과 더하여 메모리 낭비도 줄여주고 다른 Python 자료형으로도 변환이 가능한 편리한 클래스다, 라고 요약할 수 있다.

#### Exercises

1. Run the code in this section. Change the conditional statement X == Y to X < Y or X > Y, and then see what kind of tensor you can get.

In [296]:
X = tf.reshape(tf.range(12, dtype=tf.float32), (3,4))
Y = tf.constant([[2.0, 1, 4, 3], [1, 2, 3, 4], [4,3,2,1]])

X < Y, X > Y

(<tf.Tensor: shape=(3, 4), dtype=bool, numpy=
 array([[ True, False,  True, False],
        [False, False, False, False],
        [False, False, False, False]])>,
 <tf.Tensor: shape=(3, 4), dtype=bool, numpy=
 array([[False, False, False, False],
        [ True,  True,  True,  True],
        [ True,  True,  True,  True]])>)

위와 같이 == 뿐만 아니라 >, < 로도 연산이 정상적으로 되는 것을 확인할 수 있다. 위에서 != 의 경우에 대해서도 다뤘다.

2. Replace the two tensors that operate by element in the broadcasting mechanism with other shapes, e.g., 3-dimensional tensors. Is the result the same as expected?

In [297]:
a = tf.reshape(tf.range(12, dtype=tf.float32), (3,4,1))
b = tf.random.normal(shape=[1,4])

a, b, a * b

(<tf.Tensor: shape=(3, 4, 1), dtype=float32, numpy=
 array([[[ 0.],
         [ 1.],
         [ 2.],
         [ 3.]],
 
        [[ 4.],
         [ 5.],
         [ 6.],
         [ 7.]],
 
        [[ 8.],
         [ 9.],
         [10.],
         [11.]]], dtype=float32)>,
 <tf.Tensor: shape=(1, 4), dtype=float32, numpy=
 array([[ 0.3519747 ,  0.5773096 , -0.2874782 , -0.55213994]],
       dtype=float32)>,
 <tf.Tensor: shape=(3, 4, 4), dtype=float32, numpy=
 array([[[ 0.        ,  0.        , -0.        , -0.        ],
         [ 0.3519747 ,  0.5773096 , -0.2874782 , -0.55213994],
         [ 0.7039494 ,  1.1546192 , -0.5749564 , -1.1042799 ],
         [ 1.055924  ,  1.7319288 , -0.8624346 , -1.6564198 ]],
 
        [[ 1.4078988 ,  2.3092384 , -1.1499128 , -2.2085598 ],
         [ 1.7598735 ,  2.886548  , -1.437391  , -2.7606997 ],
         [ 2.111848  ,  3.4638577 , -1.7248693 , -3.3128395 ],
         [ 2.4638228 ,  4.0411673 , -2.0123475 , -3.8649795 ]],
 
        [[ 2.8157976 ,  4.618477 

위 코드는 $3 \times 4 \times 1$ Tensor인 a와 $1 \times 4$ Tensor인 b에 대해 a * b를 수행한 결과이다. 예상한 대로 shape 는 $3 \times 4 \times 4$ 가 나왔고, 제대로 Broadcasting이 된 것을 확인할 수 있었다.

## 2.2. **Data Preprocessing**

#### 2.2.1. Reading the Dataset

In [298]:
import os

os.makedirs(os.path.join('..', 'data'), exist_ok=True)
data_file = os.path.join('..', 'data', 'house_tiny.csv')
with open(data_file, 'w') as f:
    f.write('''NumRooms,RoofType,Price
NA,NA,127500
2,NA,106000
4,Slate,178100
NA,NA,140000''')

CSV는 Comma-separated values의 약자로, 표 형식의 데이터, 즉 spreadsheet와 유사한 데이터를 저장할 때 쓰인다. pandas 패키지를 이용하여 csv 파일을 handle할 수 있다.

In [299]:
import pandas as pd

data = pd.read_csv(data_file)
print(data)

   NumRooms RoofType   Price
0       NaN      NaN  127500
1       2.0      NaN  106000
2       4.0    Slate  178100
3       NaN      NaN  140000


pandas를 이용해 csv 파일을 불러오고 표시하는 코드이다.

#### 2.2.2. Data Preparation

지도 학습에서는 데이터에 라벨을 부여하여 머신러닝이 이 라벨을 토대로 학습하도록 한다. pandas의 iloc으로 column을 구분하여 input과 라벨 데이터를 준비할 수 있다.

In [300]:
inputs, targets = data.iloc[:, 0:2], data.iloc[:, 2]
inputs = pd.get_dummies(inputs, dummy_na=True)
print(inputs)
print(targets)

   NumRooms  RoofType_Slate  RoofType_nan
0       NaN               0             1
1       2.0               0             1
2       4.0               1             0
3       NaN               0             1
0    127500
1    106000
2    178100
3    140000
Name: Price, dtype: int64


#### 2.2.3. Conversion to the Tensor Format

In [301]:
import tensorflow as tf

X, y = tf.constant(inputs.values), tf.constant(targets.values)
X, y

(<tf.Tensor: shape=(4, 3), dtype=float64, numpy=
 array([[nan,  0.,  1.],
        [ 2.,  0.,  1.],
        [ 4.,  1.,  0.],
        [nan,  0.,  1.]])>,
 <tf.Tensor: shape=(4,), dtype=int64, numpy=array([127500, 106000, 178100, 140000])>)

모든 원소가 numeric하므로, Tensor로 변환할 수 있다. NaN 필드는 Tensor에서 nan으로 표현된다.

#### Exercises

1. Try loading datasets, e.g., Abalone from the UCI Machine Learning Repository and inspect their properties. What fraction of them has missing values? What fraction of the variables is numerical, categorical, or text?

In [302]:
data_file = os.path.join('drive', 'MyDrive', 'Colab Notebooks', 'Datasets', 'abalone.data')
data = pd.read_csv(data_file, names=['Sex','Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight', 'Shell weight', 'Rings'])
print(data)
print(data.info())

     Sex  Length  Diameter  Height  Whole weight  Shucked weight  \
0      M   0.455     0.365   0.095        0.5140          0.2245   
1      M   0.350     0.265   0.090        0.2255          0.0995   
2      F   0.530     0.420   0.135        0.6770          0.2565   
3      M   0.440     0.365   0.125        0.5160          0.2155   
4      I   0.330     0.255   0.080        0.2050          0.0895   
...   ..     ...       ...     ...           ...             ...   
4172   F   0.565     0.450   0.165        0.8870          0.3700   
4173   M   0.590     0.440   0.135        0.9660          0.4390   
4174   M   0.600     0.475   0.205        1.1760          0.5255   
4175   F   0.625     0.485   0.150        1.0945          0.5310   
4176   M   0.710     0.555   0.195        1.9485          0.9455   

      Viscera weight  Shell weight  Rings  
0             0.1010        0.1500     15  
1             0.0485        0.0700      7  
2             0.1415        0.2100      9  
3      

위와 같이 값이 빠진 column은 없는 것으로 확인되고, 성별을 나타내는 Sex는 M과 F, 그리고 Infant인 I로 구분되는 categorical이라 할 수 있고, 나머지는 numeric임을 확인할 수 있다.

2. Try out indexing and selecting data columns by name rather than by column number. The pandas documentation on indexing has further details on how to do this.

In [303]:
print(data['Whole weight'])

0       0.5140
1       0.2255
2       0.6770
3       0.5160
4       0.2050
         ...  
4172    0.8870
4173    0.9660
4174    1.1760
4175    1.0945
4176    1.9485
Name: Whole weight, Length: 4177, dtype: float64


위와 같이 Dictionary에 접근하는 방식으로 column에 접근할 수 있다. Indexing은 1번에서 했던 것처럼 데이터를 불러올 때 name argument에 값들을 passing함으로써 Indexing할 수 있다.

3. How large a dataset do you think you could load this way? What might be the limitations? Hint: consider the time to read the data, representation, processing, and memory footprint. Try this out on your laptop. What changes if you try it out on a server?

시스템 환경의 메모리 등의 한계가 있을 수 있다. 현재 내 데스크탑의 메모리 용량은 16 GB인데, 16 GB를 넘어가면 디스크를 이용하여 페이지를 왔다갔다 시켜야 하므로 속도가 느려질 것이다. 데스크탑보다 서버에서 하면 널널한 메모리 덕에 좀 더 큰 데이터를 무리없이 로드할 수 있을 것이다.

4. How would you deal with data that has a very large number of categories? What if the category labels are all unique? Should you include the latter?

만약 데이터마다 카데고리의 종류가 매우 크다면, 나는 이 column을 포함시킬 지 고민할 것 같다. 학습에 있어서 별다른 도움이 되지 않을 것 같기 때문이다. 만약 카테고리가 모두 다르다면, 나는 포함시키지 않을 것이다.

5. What alternatives to pandas can you think of? How about loading NumPy tensors from a file? Check out Pillow, the Python Imaging Library.

NumPy의 load 메소드를 이용하는 것도 대체제가 될 수 있겠다. 하지만 numeric data에 한해서만 가능할 것이다. PIL을 이용하면 이미지 데이터를 학습에 이용할 수 있을 것이다.

## 2.3. **Linear Algebra**

#### 2.3.1. Scalars

In [304]:
x = tf.constant(3.0)
y = tf.constant(2.0)
z = tf.constant([3.0])
x + y, x * y, x / y, x**y, z

(<tf.Tensor: shape=(), dtype=float32, numpy=5.0>,
 <tf.Tensor: shape=(), dtype=float32, numpy=6.0>,
 <tf.Tensor: shape=(), dtype=float32, numpy=1.5>,
 <tf.Tensor: shape=(), dtype=float32, numpy=9.0>,
 <tf.Tensor: shape=(1,), dtype=float32, numpy=array([3.], dtype=float32)>)

위와 같이 list 형태가 아닌, numeric value를 argument로 주면 scalar를 생성할 수 있다. scalar는 shape가 z와 다르게게 ()이다.

#### 2.3.2. Vectors

In [305]:
x = tf.range(3)
y = tf.random.normal([4])
z = tf.constant([2,56,3.0])
x,y,z

(<tf.Tensor: shape=(3,), dtype=int32, numpy=array([0, 1, 2], dtype=int32)>,
 <tf.Tensor: shape=(4,), dtype=float32, numpy=array([-0.01451359,  0.6978944 ,  0.9672288 ,  0.38913828], dtype=float32)>,
 <tf.Tensor: shape=(3,), dtype=float32, numpy=array([ 2., 56.,  3.], dtype=float32)>)

위와 같이 다양한 방법으로 벡터를 생성할 수 있다. 스칼라는 0<sup>th</sup>-order Tensor이고, 벡터는 1<sup>st</sup>-order Tensor이다. 뒷 절에서 다루는 행렬은 2<sup>nd</sup>-order Tensor이다.

#### 2.3.3. Matrices

In [306]:
A = tf.reshape(tf.range(6), (3, 2))
B = tf.random.normal([3,2])
C = tf.constant([[1,3,7],[2,7,2]])
A,B,C

(<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[0, 1],
        [2, 3],
        [4, 5]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.5803961 , -1.1074721 ],
        [-0.7726394 ,  1.1891983 ],
        [ 0.23741618, -0.29158148]], dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[1, 3, 7],
        [2, 7, 2]], dtype=int32)>)

벡터와 유사하게 다양한 방법으로 행렬을 생성할 수 있다. 위에서 언급했듯 행렬은 2<sup>nd</sup>-order Tensor이다.

In [307]:
tf.transpose(A)

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[0, 2, 4],
       [1, 3, 5]], dtype=int32)>

전치 연산은 transpose 메소드로 가능하다.

In [308]:
A = tf.constant([[1, 2, 3], [2, 0, 4], [3, 4, 5]])
A == tf.transpose(A)

<tf.Tensor: shape=(3, 3), dtype=bool, numpy=
array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])>

대칭행렬은 전치한 행렬과 같기 때문에 모든 원소가 True인 모습이다.

#### 2.3.4. Tensors

In [309]:
A = tf.reshape(tf.range(24), (2, 3, 4))
B = tf.random.normal([3,2,4])
C = tf.constant([[[1,3,7],[2,7,2]],[[5,2,7],[1,2,9]]])
A,B,C

(<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
 array([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],
 
        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2, 4), dtype=float32, numpy=
 array([[[-1.3035291 , -1.2911451 , -1.2039193 ,  1.2929406 ],
         [-1.6534953 ,  0.6223284 ,  0.6293017 ,  0.40589404]],
 
        [[ 2.0652878 , -2.1153324 ,  1.2286118 ,  1.1994016 ],
         [-0.26548886, -0.05183019, -0.7039975 , -1.2902461 ]],
 
        [[-0.51726425,  0.61205876,  0.7766286 ,  1.0722406 ],
         [ 0.9057584 , -0.02201245,  0.08268177,  0.85874563]]],
       dtype=float32)>,
 <tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
 array([[[1, 3, 7],
         [2, 7, 2]],
 
        [[5, 2, 7],
         [1, 2, 9]]], dtype=int32)>)

역시 다양한 방법으로 텐서를 생성할 수 있다. 스칼라도 텐서이고, 벡터도 텐서이고, 행렬도 텐서이다. 텐서는 저 셋을 모두 포괄하여, 더 나아가서 일반적인 형식으로 이해하였다. 스칼라든, 벡터든, 행렬이든, 텐서든 생성 방식이 다 똑같다는 점을 알 수 있다.

#### 2.3.5. Basic Properties of Tensor Arithmetic

In [310]:
A = tf.reshape(tf.range(6, dtype=tf.float32), (2, 3))
B = A  # No cloning of A to B by allocating new memory
A, A + B

(<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[0., 1., 2.],
        [3., 4., 5.]], dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[ 0.,  2.,  4.],
        [ 6.,  8., 10.]], dtype=float32)>)

위는 (2,3) shape의 텐서를 생성하고, 같은 두 텐서를 더한 코드이다.

In [311]:
A * B

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[ 0.,  1.,  4.],
       [ 9., 16., 25.]], dtype=float32)>

위는 행렬곱과는 다른, Element-wise product를 수행하는 Hadamard Product이다. Hadamard Product는 $\odot$으로 표현한다.

In [312]:
a = 2
X = tf.reshape(tf.range(24), (2, 3, 4))
a + X, (a * X)

(<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
 array([[[ 2,  3,  4,  5],
         [ 6,  7,  8,  9],
         [10, 11, 12, 13]],
 
        [[14, 15, 16, 17],
         [18, 19, 20, 21],
         [22, 23, 24, 25]]], dtype=int32)>,
 <tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
 array([[[ 0,  2,  4,  6],
         [ 8, 10, 12, 14],
         [16, 18, 20, 22]],
 
        [[24, 26, 28, 30],
         [32, 34, 36, 38],
         [40, 42, 44, 46]]], dtype=int32)>)

Broadcasting을 적용하여 스칼라와 텐서를 각각 더하고 곱한 코드이다. 덧셈의 경우 X와 차원이 같고 모든 원소가 2인 텐서를 더한 것이고, 곱셈은 상수배를 한 것이다. 같은 차원이 결과로 나온 것으로 보아, 텐서도 벡터 공간과 같이 텐서 공간을 이룬다는 것으로 이해하였다. 이에 대해 좀 더 찾아보니, 텐서공간은 텐서덧셈과 텐서곱이 정의된 공간이라고 한다. 즉 저 코드에서 곱셈은 X와 차원이 같고 모든 원소가 2인 텐서를 Hadamard Product를 수행한 것이다.

#### 2.3.6. Reduction

In [313]:
x = tf.range(3, dtype=tf.float32)
x, tf.reduce_sum(x)

(<tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 1., 2.], dtype=float32)>,
 <tf.Tensor: shape=(), dtype=float32, numpy=3.0>)

2.1절에서 본 것과 같이 reduce_sum 메소드로 모든 원소의 합을 구할 수 있다. 결과는 스칼라로 나온다.

In [314]:
A.shape, tf.reduce_sum(A)

(TensorShape([2, 3]), <tf.Tensor: shape=(), dtype=float32, numpy=15.0>)

In [315]:
A.shape, tf.reduce_sum(A, axis=0)

(TensorShape([2, 3]),
 <tf.Tensor: shape=(3,), dtype=float32, numpy=array([3., 5., 7.], dtype=float32)>)

In [316]:
A.shape, tf.reduce_sum(A, axis=1)

(TensorShape([2, 3]),
 <tf.Tensor: shape=(2,), dtype=float32, numpy=array([ 3., 12.], dtype=float32)>)

Axis argument를 주어 축 별로 합을 계산할 수 있다.

In [317]:
tf.reduce_sum(A, axis=[0, 1]), tf.reduce_sum(A)

(<tf.Tensor: shape=(), dtype=float32, numpy=15.0>,
 <tf.Tensor: shape=(), dtype=float32, numpy=15.0>)

위와 같이 모든 축에 대한 합을 구하면 그냥 행렬만 넘겨주는 것과 결과가 같음을 볼 수 있다.

In [318]:
tf.reduce_mean(A), tf.reduce_sum(A) / tf.size(A).numpy()

(<tf.Tensor: shape=(), dtype=float32, numpy=2.5>,
 <tf.Tensor: shape=(), dtype=float32, numpy=2.5>)

In [319]:
tf.reduce_mean(A, axis=0), tf.reduce_sum(A, axis=0) / A.shape[0]

(<tf.Tensor: shape=(3,), dtype=float32, numpy=array([1.5, 2.5, 3.5], dtype=float32)>,
 <tf.Tensor: shape=(3,), dtype=float32, numpy=array([1.5, 2.5, 3.5], dtype=float32)>)

In [320]:
tf.reduce_mean(A, axis=1), tf.reduce_sum(A, axis=1) / A.shape[1]

(<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 4.], dtype=float32)>,
 <tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 4.], dtype=float32)>)

reduce_mean 메소드로 reduce_sum과 유사하게 축별로 평균을 구할 수 있다.

#### 2.3.7. Non-Reduction Sum

In [321]:
sum_A = tf.reduce_sum(A, axis=1, keepdims=True)
sum_A, sum_A.shape

(<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
 array([[ 3.],
        [12.]], dtype=float32)>,
 TensorShape([2, 1]))

keepdims argument를 이용하여 차원을 유지한 채로 합을 구할 수 있다. 물론 합을 구하면 해당하는 축의 차원은 1이 될 것이다.

In [322]:
A / sum_A

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[0.        , 0.33333334, 0.6666667 ],
       [0.25      , 0.33333334, 0.41666666]], dtype=float32)>

sum_A를 이용하여 axis 1에 해당하는 각 원소의 합을 1로 만드는 코드이다. axis를 1로 설정하였으므로, 안쪽 리스트 당 그 원소의 합이 1임을 볼 수 있다.

In [323]:
tf.cumsum(A, axis=0)

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[0., 1., 2.],
       [3., 5., 7.]], dtype=float32)>

In [324]:
tf.cumsum(A, axis=1)

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[ 0.,  1.,  3.],
       [ 3.,  7., 12.]], dtype=float32)>

위 코드는 cumsum 메소드로 축을 따라 누적합을 계산하는 코드이다. axis가 0일 때는 행을 따라 누적합을 구하였고, axis가 1일 때는 열을 따라 누적합을 구한 모습이다.

#### 2.3.8 Dot Products

In [325]:
y = tf.ones(3, dtype=tf.float32)
x, y, tf.tensordot(x, y, axes=1)

(<tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 1., 2.], dtype=float32)>,
 <tf.Tensor: shape=(3,), dtype=float32, numpy=array([1., 1., 1.], dtype=float32)>,
 <tf.Tensor: shape=(), dtype=float32, numpy=3.0>)

In [326]:
tf.reduce_sum(x * y)

<tf.Tensor: shape=(), dtype=float32, numpy=3.0>

tensordot 메소드를 이용해 내적을 구할 수 있다.

#### 2.3.9. Matrix-Vector Products

In [327]:
A, x, tf.linalg.matvec(A, x)

(<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[0., 1., 2.],
        [3., 4., 5.]], dtype=float32)>,
 <tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 1., 2.], dtype=float32)>,
 <tf.Tensor: shape=(2,), dtype=float32, numpy=array([ 5., 14.], dtype=float32)>)

linalg 클래스의 matvec 메소드를 사용하여 행렬과 벡터의 곱을 계산할 수 있다.

#### 2.3.10. Matrix-Matrix Multiplication

In [328]:
B = tf.reshape(tf.range(12, dtype=tf.float32), shape=(3,4))
A, B, tf.matmul(A, B)

(<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[0., 1., 2.],
        [3., 4., 5.]], dtype=float32)>,
 <tf.Tensor: shape=(3, 4), dtype=float32, numpy=
 array([[ 0.,  1.,  2.,  3.],
        [ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.]], dtype=float32)>,
 <tf.Tensor: shape=(2, 4), dtype=float32, numpy=
 array([[20., 23., 26., 29.],
        [56., 68., 80., 92.]], dtype=float32)>)

matmul 메소드를 사용하여 행렬곱도 구할 수 있다.

#### 2.3.11. Norms

In [329]:
tf.norm(tf.ones((4, 9)))

<tf.Tensor: shape=(), dtype=float32, numpy=6.0>

norm 메소드를 사용하여 텐서의 Norm을 구할 수 있다.

#### Exercises

이 절에서는 코딩과 관련된 문제를 선별하여 풀었다.

4. We defined the tensor X of shape (2, 3, 4) in this section. What is the output of len(X)? Write your answer without implementing any code, then check your answer using code.

In [330]:
X = tf.reshape(tf.range(24, dtype=tf.float32), shape=(2,3,4))
len(X)

2

길이를 뜻하는 len(X)가 2가 나왔다. 여기서 나는 가장 바깥 axis를 기준으로 길이를 계산한다고 생각했다.

5. For a tensor X of arbitrary shape, does len(X) always correspond to the length of a certain axis of X? What is that axis?

In [331]:
X = tf.reshape(tf.range(24, dtype=tf.float32), shape=(4,2,3))
Y = tf.reshape(tf.range(24, dtype=tf.float32), shape=(3,4,2))
len(X), len(Y)

(4, 3)

예측이 맞은 것 같다.

6. Run A / A.sum(axis=1) and see what happens. Can you analyze the reason?

In [332]:
A = tf.reshape(tf.range(6, dtype=tf.float32), shape=(2,3))
A, tf.reduce_sum(A,axis=1, keepdims=True), A / tf.reduce_sum(A,axis=1, keepdims=True)

(<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[0., 1., 2.],
        [3., 4., 5.]], dtype=float32)>,
 <tf.Tensor: shape=(2, 1), dtype=float32, numpy=
 array([[ 3.],
        [12.]], dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[0.        , 0.33333334, 0.6666667 ],
        [0.25      , 0.33333334, 0.41666666]], dtype=float32)>)

axis를 1로 설정하면 열마다 sum을 수행하여 shape가 (2,)가 나오게 되므로 Broadcasting이 불가능하다. 이럴 때에는 keepdims를 True로 설정하여 Broadcasting이 가능하도록 해주면 된다.

In [333]:
A = tf.reshape(tf.range(6, dtype=tf.float32), shape=(2,3))
A / tf.reduce_sum(A,axis=0)

<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[0.        , 0.2       , 0.2857143 ],
       [1.        , 0.8       , 0.71428573]], dtype=float32)>

추가로 axis를 0으로 설정하여 실행한 결과이다.

8. Consider a tensor with shape (2, 3, 4). What are the shapes of the summation outputs along axis 0, 1, and 2?

In [334]:
X = tf.reshape(tf.range(24,dtype=tf.float32), shape=(2,3,4))
X, tf.reduce_sum(X, axis=0),tf.reduce_sum(X, axis=1),tf.reduce_sum(X, axis=2),

(<tf.Tensor: shape=(2, 3, 4), dtype=float32, numpy=
 array([[[ 0.,  1.,  2.,  3.],
         [ 4.,  5.,  6.,  7.],
         [ 8.,  9., 10., 11.]],
 
        [[12., 13., 14., 15.],
         [16., 17., 18., 19.],
         [20., 21., 22., 23.]]], dtype=float32)>,
 <tf.Tensor: shape=(3, 4), dtype=float32, numpy=
 array([[12., 14., 16., 18.],
        [20., 22., 24., 26.],
        [28., 30., 32., 34.]], dtype=float32)>,
 <tf.Tensor: shape=(2, 4), dtype=float32, numpy=
 array([[12., 15., 18., 21.],
        [48., 51., 54., 57.]], dtype=float32)>,
 <tf.Tensor: shape=(2, 3), dtype=float32, numpy=
 array([[ 6., 22., 38.],
        [54., 70., 86.]], dtype=float32)>)

결과는 위와 같다. 각 축에 해당하는 차원이 각 결과마다 없어진 것을 볼 수 있다. 기준이 되었기 때문이다.

9. Feed a tensor with 3 or more axes to the linalg.norm function and observe its output. What does this function compute for tensors of arbitrary shape?

In [335]:
X = tf.reshape(tf.range(48,dtype=tf.float32), shape=(2,3,2,4))
X, tf.norm(X)

(<tf.Tensor: shape=(2, 3, 2, 4), dtype=float32, numpy=
 array([[[[ 0.,  1.,  2.,  3.],
          [ 4.,  5.,  6.,  7.]],
 
         [[ 8.,  9., 10., 11.],
          [12., 13., 14., 15.]],
 
         [[16., 17., 18., 19.],
          [20., 21., 22., 23.]]],
 
 
        [[[24., 25., 26., 27.],
          [28., 29., 30., 31.]],
 
         [[32., 33., 34., 35.],
          [36., 37., 38., 39.]],
 
         [[40., 41., 42., 43.],
          [44., 45., 46., 47.]]]], dtype=float32)>,
 <tf.Tensor: shape=(), dtype=float32, numpy=188.99736>)

Norm의 정의는 각 원소를 제곱한 합의 제곱근이므로 결과는 sqrt(35720) = 188.99736...가 나온 것을 확인할 수 있다.

10. Define three large matrices, say $A \in ℝ$<sup>2<sup>10</sup>$\times$ 2<sup>16</sup></sup>, $B \in ℝ$<sup>2<sup>16</sup>$\times$ 2<sup>5</sup></sup> and $C \in ℝ$<sup>2<sup>5</sup>$\times$ 2<sup>14</sup></sup>, for instance initialized with Gaussian random variables. You want to compute the product $ABC$.
Is there any difference in memory footprint and speed, depending on whether you compute $(AB)C$ or $A(BC)$. Why?

$(AB)C$는 먼저 $AB$를 계산할 때 2<sup>10</sup>$\times$ 2<sup>5</sup> $\times$ 2<sup>16</sup>번의 곱연산, 그 후 2<sup>10</sup> $\times$ 2<sup>14</sup> $\times$ 2<sup>5</sup>번의 곱연산을 한다. $A(BC)$는 먼저 $BC$를 계산할 때 2<sup>16</sup> $\times$ 2<sup>14</sup> $\times$ 2<sup>5</sup>번의 곱연산, 그 후 2<sup>10</sup> $\times$ 2<sup>14</sup> $\times$ 2<sup>16</sup> 번의 곱연산을 한다. 전자의 경우는 총 5 $\times$ 2<sup>29</sup>번의 곱연산을 하고, 후자의 경우는 33 $\times$ 2<sup>35</sup>번의 곱연산을 하므로 계산 시간이 약 420배 차이가 날 것으로 예측된다. 전자가 훨씬 더 빠르다. 메모리도 전자의 경우가 덜 잡아먹는다.

11. Define three large matrices, say $A \in ℝ$<sup>2<sup>10</sup>$\times$ 2<sup>16</sup></sup>, $B \in ℝ$<sup>2<sup>16</sup>$\times$ 2<sup>5</sup></sup> and $C \in ℝ$<sup>2<sup>5</sup>$\times$ 2<sup>16</sup></sup>. Is there any difference in speed depending on whether you compute $AB$ or $AC$<sup>T</sup>? Why? What changes if you initialize $C = B$<sup>T</sup> without cloning memory? Why?

transpose를 할 때 메모리 소비 및 전치행렬을 만드는 데 걸리는 시간으로 인해 $AB$를 계산하는 쪽이 빠를 것 같다. 만약 메모리를 할당하여 전치행렬을 만드는 과정없이 $C = B$<sup>T</sup>으로 정의한다면 속도는 유사할 것으로 예측된다. 전치의 전치는 원래 행렬이기 때문이다. Tensorflow에서 최적화가 이런 방법으로 되었을 지는 의문이지만. NumPy에는 View라는 개념이 있는데, Tensorflow도 Tensor 내부를 까보면 NumPy 클래스가 구성요소에 있으므로 혹시 모른다.

12. Define three matrices, say $A, B, C \in ℝ^{100\times200} $. Constitute a tensor with 3 axes by stacking 
$[A, B, C]$. What is the dimensionality? Slice out the second coordinate of the third axis to recover 
$B$. Check that your answer is correct.

In [336]:
A = tf.random.normal([100,200])
B = tf.random.normal([100,200])
C = tf.random.normal([100,200])
D = tf.reshape(tf.concat([A,B,C], axis=0), shape=(3,100,200))
B, D[1:2]

(<tf.Tensor: shape=(100, 200), dtype=float32, numpy=
 array([[-0.7340673 , -0.29955998,  0.02083631, ...,  0.41261768,
         -0.83056283, -1.4059781 ],
        [ 0.03017859, -0.63480926, -0.25602758, ...,  0.47184876,
          0.9594048 ,  0.28535533],
        [-0.3828047 , -0.60594845, -0.74002403, ...,  1.0304345 ,
         -0.66154474, -0.02317384],
        ...,
        [-0.9389598 ,  0.47585806,  0.15916027, ...,  0.7364893 ,
          0.01138927, -2.099889  ],
        [-0.27575135, -0.77013534,  0.09127738, ..., -0.2393353 ,
          0.8665059 ,  0.39878196],
        [-1.2032483 ,  1.2952024 , -0.2175503 , ...,  2.0283172 ,
         -0.74139005,  0.93545014]], dtype=float32)>,
 <tf.Tensor: shape=(1, 100, 200), dtype=float32, numpy=
 array([[[-0.7340673 , -0.29955998,  0.02083631, ...,  0.41261768,
          -0.83056283, -1.4059781 ],
         [ 0.03017859, -0.63480926, -0.25602758, ...,  0.47184876,
           0.9594048 ,  0.28535533],
         [-0.3828047 , -0.60594845, -0.7

First Axis로 자르는게 맞지 않나 싶다...

## 2.4. **Calculus**

In [337]:
%matplotlib inline
import numpy as np
from matplotlib_inline import backend_inline
from d2l import tensorflow as d2l

#### 2.4.1. Derivatives and Differentiation

In [338]:
def f(x):
    return 3 * x ** 2 - 4 * x

for h in 10.0**np.arange(-1, -6, -1):
    print(f'h={h:.5f}, numerical limit={(f(1+h)-f(1))/h:.5f}')

h=0.10000, numerical limit=2.30000
h=0.01000, numerical limit=2.03000
h=0.00100, numerical limit=2.00300
h=0.00010, numerical limit=2.00030
h=0.00001, numerical limit=2.00003


다음은 h값들에 대해 수치미분을 한 결과이다. h가 0에 가까울 수록 해석미분의 값과 가까워진다.

#### 2.4.2. Visualization Utilities

matplotlib 패키지를 통해 함수 개형과 접선 등등을 시각화할 수 있다.

(#@save 주석이 코랩에서 지원되지 않는 것 같다. #@save는 d2l 패키지에서 반복적인 코드 작성을 피하게끔 해주지만 여기서는 #@save 사용 없이 써야 할 것 같다.)

In [339]:
def use_svg_display():
    """Use the svg format to display a plot in Jupyter."""
    backend_inline.set_matplotlib_formats('svg')

def set_figsize(figsize=(3.5, 2.5)):
    """Set the figure size for matplotlib."""
    use_svg_display()
    d2l.plt.rcParams['figure.figsize'] = figsize

def set_axes(axes, xlabel, ylabel, xlim, ylim, xscale, yscale, legend):
    """Set the axes for matplotlib."""
    axes.set_xlabel(xlabel), axes.set_ylabel(ylabel)
    axes.set_xscale(xscale), axes.set_yscale(yscale)
    axes.set_xlim(xlim),     axes.set_ylim(ylim)
    if legend:
        axes.legend(legend)
    axes.grid()

def plot(X, Y=None, xlabel=None, ylabel=None, legend=[], xlim=None,
         ylim=None, xscale='linear', yscale='linear',
         fmts=('-', 'm--', 'g-.', 'r:'), figsize=(3.5, 2.5), axes=None):
    """Plot data points."""

    def has_one_axis(X):  # True if X (tensor or list) has 1 axis
        return (hasattr(X, "ndim") and X.ndim == 1 or isinstance(X, list)
                and not hasattr(X[0], "__len__"))

    if has_one_axis(X): X = [X]
    if Y is None:
        X, Y = [[]] * len(X), X
    elif has_one_axis(Y):
        Y = [Y]
    if len(X) != len(Y):
        X = X * len(Y)

    set_figsize(figsize)
    if axes is None:
        axes = d2l.plt.gca()
    axes.cla()
    for x, y, fmt in zip(X, Y, fmts):
        axes.plot(x,y,fmt) if len(x) else axes.plot(y,fmt)
    set_axes(axes, xlabel, ylabel, xlim, ylim, xscale, yscale, legend)

x = np.arange(0, 3, 0.1)
plot(x, [f(x), 2 * x - 3], 'x', 'f(x)', legend=['f(x)', 'Tangent line (x=1)'])

ImportError: ignored

<Figure size 350x250 with 1 Axes>

ImportError 문제로 코드를 실행하지 못하였다... 위 코드를 요약하자면 d2l의 plt를 이용하여 그래프를 보여준다.

(이번 절에서는 코드가 별로 없고 미적분학에 관한 내용이 대부분이므로 Exercise 또학 코딩이 없고 미적분 관련 증명, 계산 문제이기 때문에 넘어가도록 하겠다...)

## 2.5. **Automatic Differentiation**

#### 2.5.1. A Simple Function

In [340]:
x = tf.range(4, dtype=tf.float32)
x

<tf.Tensor: shape=(4,), dtype=float32, numpy=array([0., 1., 2., 3.], dtype=float32)>

In [341]:
x = tf.Variable(x)
x

<tf.Variable 'Variable:0' shape=(4,) dtype=float32, numpy=array([0., 1., 2., 3.], dtype=float32)>

In [342]:
# Record all computations onto a tape
with tf.GradientTape() as t:
    y = 2 * tf.tensordot(x, x, axes=1)
y

<tf.Tensor: shape=(), dtype=float32, numpy=28.0>

In [343]:
x_grad = t.gradient(y, x)
x_grad

<tf.Tensor: shape=(4,), dtype=float32, numpy=array([ 0.,  4.,  8., 12.], dtype=float32)>

In [344]:
x_grad == 4 * x

<tf.Tensor: shape=(4,), dtype=bool, numpy=array([ True,  True,  True,  True])>

위 코드들은 벡터 $x$에 대해 $y = 2x^{T}x$에 대한 Gradient를 관측한다. 당연한 사실이지만, 마지막 줄 코드로 Gradient는 4$x$로 표현될 수 있음을 알 수 있다. 

In [345]:
with tf.GradientTape() as t:
    y = x * x
t.gradient(y, x)  # Same as y = tf.reduce_sum(x * x)

<tf.Tensor: shape=(4,), dtype=float32, numpy=array([0., 2., 4., 6.], dtype=float32)>

이 경우 Gradient는 2$x$가 되겠다.

#### 2.5.3 Detaching Computation

In [346]:
# Set persistent=True to preserve the compute graph.
# This lets us run t.gradient more than once
with tf.GradientTape(persistent=True) as t:
    y = x * x
    u = tf.stop_gradient(y)
    z = u * x

x_grad = t.gradient(z, x)
x_grad == u

<tf.Tensor: shape=(4,), dtype=bool, numpy=array([ True,  True,  True,  True])>

위 코드는 편미분과 언뜻 비슷해 보이지만, 약간 다른 기능에 대해 보여준다. 원래는 $y$가 $x$에 관한 식이기 때문에 $y$를 고정하고 $x$에 대한 $z$의 편미분을 구한다는 것은 이상해 보이지만, 위 코드는 실제로 그렇게 한다. $y$를 stop_gradient를 통해 $u$라는 어떤 고정값을 두고, 거기서 $x$ 에 대한 편미분을 구한다. 이 때 $z$는 $ux$로 나타내기 때문에 편미분은 $x$가 된다. d2l의 설명을 보면 "we can create a new variable u that takes the same value as y but whose provenance (how it was created) has been wiped out." 라고 설명한다. $y$가 어떤 변수에 관한 식인지에 대한한 그런 걸 싹 다 무시한다고 나와있다.

In [347]:
t.gradient(y, x) == 2 * x

<tf.Tensor: shape=(4,), dtype=bool, numpy=array([ True,  True,  True,  True])>

$y$에 대해서 여전히 정상적으로 $2x$로 나타나는 것을 볼 수 있다.

#### 2.5.4. Gradients and Python Control Flow

In [348]:
def f(a):
    b = a * 2
    while tf.norm(b) < 1000:
        b = b * 2
    if tf.reduce_sum(b) > 0:
        c = b
    else:
        c = 100 * b
    return c

a = tf.Variable(tf.random.normal(shape=()))
with tf.GradientTape() as t:
    d = f(a)
d_grad = t.gradient(d, a)
d, a, d_grad

(<tf.Tensor: shape=(), dtype=float32, numpy=1619.703>,
 <tf.Variable 'Variable:0' shape=() dtype=float32, numpy=1.5817412>,
 <tf.Tensor: shape=(), dtype=float32, numpy=1024.0>)

In [349]:
d_grad == d / a

<tf.Tensor: shape=(), dtype=bool, numpy=True>

위 예시는 조건문이나 반복문 등 Python의 Control Flow가 포함된 함수에서도 Gradient를 구할 수 있다는 것을 보여주는 코드이다. f를 보면, 반복문과 조건문이 포함되어 있다. 언뜻 보면 Gradient를 구하는 것이 까다로워 보이지만, 결국 f는 a에 대한 선형식이기 때문에, f(a)/a가 Gradient가 된다. 코드에서 알 수 있듯이 f처럼 복잡한 함수에 대해 gradient 메소드로 Gradient를 간단히 구한 것과 같다. 딥러닝에서 Control Flow가 밥먹듯이 많이 쓰이기 때문에, 이러한 기능은 중요하다.

(이 절의 Exercise도 대부분 수학 관련이기 때문에 넘어가겠다.)

## 2.6. **Probability and Statistics**

In [350]:
%matplotlib inline
import random
import tensorflow as tf
from tensorflow_probability import distributions as tfd
from d2l import tensorflow as d2l

#### 2.6.1. A Simple Example: Tossing Coins

In [351]:
num_tosses = 100
heads = sum([random.random() > 0.5 for _ in range(100)])
tails = num_tosses - heads
print("heads, tails: ", [heads, tails])

heads, tails:  [45, 55]


확률의 단순한 예로 동전 던지기를 구현한 코드이다.

In [352]:
fair_probs = tf.ones(2) / 2
tfd.Multinomial(100, fair_probs).sample()

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([49., 51.], dtype=float32)>

In [353]:
tfd.Multinomial(100, fair_probs).sample() / 100

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([0.53, 0.47], dtype=float32)>

다항분포를 이용한 좀 더 일반화된 코드이다.

In [354]:
counts = tfd.Multinomial(10000, fair_probs).sample()
counts / 10000

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([0.4958, 0.5042], dtype=float32)>

표본이 많을수록 오차가 적음을 보여주는 코드이다. 오차가 ${1} \over \sqrt{n}$에 비례한다는 것은 잘 알려진 사실이다.

In [355]:
counts = tfd.Multinomial(1, fair_probs).sample(10000)
cum_counts = tf.cumsum(counts, axis=0)
estimates = cum_counts / tf.reduce_sum(cum_counts, axis=1, keepdims=True)
estimates = estimates.numpy()

d2l.set_figsize((4.5, 3.5))
d2l.plt.plot(estimates[:, 0], label=("P(coin=heads)"))
d2l.plt.plot(estimates[:, 1], label=("P(coin=tails)"))
d2l.plt.axhline(y=0.5, color='black', linestyle='dashed')
d2l.plt.gca().set_xlabel('Samples')
d2l.plt.gca().set_ylabel('Estimated probability')
d2l.plt.legend();

ImportError: ignored

<Figure size 450x350 with 1 Axes>

2.5절에서의 ImportError와 똑같은 에러가 발생하였다... 뭔가 d2l 패키지와 matplotlib이 호환이 안되는 것 같다. 요약하자면 오차가 위 예시에서 보인 식을 따라간다는 것을 시각화해주는 코드이다.

이 절에서 실습할 코드는 남아있지 않고 남은 내용은 확률과 통계 관련 내용이니 내용에 대해 짧게 요약하겠다.

2.6.3 절은 확률변수에 대해 기본적인 내용을 다룬다.

2.6.4 절은 다변수 확률변수와 베이즈 정리에 대해 다룬다.

2.6.5 절은 기댓값과 분산, 공분산에 대해 다룬다.

(이 절의 Exercise 역시 계산 관련 문제이므로 넘어가도록 하겠다.)

## 2.7. **Documentation**

이 절에서는 프레임워크 메소드에 대한 도움말이나 관련 정보를 얻을 수 있는 팁에 대해 짤막하게 소개한다.

In [356]:
print(dir(tf.random))

['Algorithm', 'Generator', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '_sys', 'all_candidate_sampler', 'categorical', 'create_rng_state', 'experimental', 'fixed_unigram_candidate_sampler', 'fold_in', 'gamma', 'get_global_generator', 'learned_unigram_candidate_sampler', 'log_uniform_candidate_sampler', 'normal', 'poisson', 'set_global_generator', 'set_seed', 'shuffle', 'split', 'stateless_binomial', 'stateless_categorical', 'stateless_gamma', 'stateless_normal', 'stateless_parameterized_truncated_normal', 'stateless_poisson', 'stateless_truncated_normal', 'stateless_uniform', 'truncated_normal', 'uniform', 'uniform_candidate_sampler']


dir를 통해 Subclass, 혹은 메소드에 대해 탐색할 수 있다.

In [357]:
help(tf.ones)

Help on function ones in module tensorflow.python.ops.array_ops:

ones(shape, dtype=tf.float32, name=None)
    Creates a tensor with all elements set to one (1).
    
    See also `tf.ones_like`, `tf.zeros`, `tf.fill`, `tf.eye`.
    
    This operation returns a tensor of type `dtype` with shape `shape` and
    all elements set to one.
    
    >>> tf.ones([3, 4], tf.int32)
    <tf.Tensor: shape=(3, 4), dtype=int32, numpy=
    array([[1, 1, 1, 1],
           [1, 1, 1, 1],
           [1, 1, 1, 1]], dtype=int32)>
    
    Args:
      shape: A `list` of integers, a `tuple` of integers, or
        a 1-D `Tensor` of type `int32`.
      dtype: Optional DType of an element in the resulting `Tensor`. Default is
        `tf.float32`.
      name: Optional string. A name for the operation.
    
    Returns:
      A `Tensor` with all elements set to one (1).



help 메소드를 통해 해당 메소드 혹은 클래스에 관한 개요를 볼 수 있다.